asf_heat_pump_suitability/analysis/exploratory/create_full_dataset/Merging_clusters_plymouth.py

```diff
@@ -0,0 +1,419 @@
+
+import os
+import random
+
+from asf_heat_pump_suitability.pipeline.prepare_features import anchor_properties
+
+# %% [markdown]
+# ## Read data for Plymouth
+# - UPRNs and which cluster they are part of
+# - Building polygons
+# - Anchor properties
+
+# %%
+UPRNs_clustered = gpd.read_file(
+    "s3://asf-heat-pump-suitability/exploration/spatial_clustering_plymouth/plymouth_residential_uprns_with_clusters/plymouth_residential_uprns_with_clusters.shp"
+).to_crs(epsg=4326)
+
+# %%
+# Instead of all none clusters being assigned the -1 cluster, change this to a unique number - use the negative of the UPRN since this will be so much bigger, so won't overlap with genuine clusters
+UPRNs_clustered["cluster"] = UPRNs_clustered.apply(
+    lambda x: x["cluster"] if x["cluster"] >= 0 else -x["UPRN"], axis=1
+)
+
+# %%
+assert (
+    (UPRNs_clustered["cluster"] < 0).sum()
+    + UPRNs_clustered[UPRNs_clustered["cluster"] >= 0]["cluster"].nunique()
+) == UPRNs_clustered["cluster"].nunique()
+
+# %%
+print("\nLoading OS OpenMap Local - All buildings in SX region...")
+os_openmap_building_gdf = gpd.read_file(
+    "s3://asf-heat-pump-suitability/exploration/spatial_clustering_plymouth/OS OpenMap Local (ESRI Shape File) SX/data/SX_Building.shp"
+).to_crs(epsg=4326)
+
+# %%
+print("\nFilter all buildings in SX region to just be in Plymouth...")
+# There is probably a better way to do this using a Plymouth polygon - but this will do for now
+building_footprints = os_openmap_building_gdf.cx[
+    UPRNs_clustered["LONGITUDE"].min() : UPRNs_clustered["LONGITUDE"].max(),
+    UPRNs_clustered["LATITUDE"].min() : UPRNs_clustered["LATITUDE"].max(),
+]
+
+# %%
+print(
+    f"There are {len(building_footprints)} buildings in Plymouth and {len(UPRNs_clustered)} UPRNs"
+)
+
+# %%
+# Load anchor properties (for distance to anchor property per cluster)
+anchor_properties_gdf = anchor_properties.load_gdf_and_process_poi()
+
+# %% [markdown]
+# ## Process and join data
+#
+# Find which building polygons each UPRN point sits within (if any)
+
+# %%
+UPRNs_clustered = gpd.GeoDataFrame(
+    UPRNs_clustered,
+    geometry=gpd.points_from_xy(UPRNs_clustered.LONGITUDE, UPRNs_clustered.LATITUDE),
+    crs="EPSG:4326",
+)
+
+# %%
+building_footprints["building_geom"] = (
+    building_footprints.geometry
+)  # Otherwise there are 2 geometry columns and this one gets dropped in the join
+
+# %%
+UPRNs_joined_buildings = gpd.sjoin(
+    UPRNs_clustered,
+    building_footprints,
+    how="inner",
+    predicate="intersects",
+)
+
+# %% [markdown]
+# Finding out a few stats about this join
+
+# %%
+print(len(building_footprints))
+print(len(UPRNs_clustered))
+print(len(UPRNs_joined_buildings))
+
+# %%
+print(
+    f"{UPRNs_joined_buildings['UPRN'].nunique()} out of {UPRNs_clustered['UPRN'].nunique()} unique UPRNs were found within building polygons. "
+)
+
+# %%
+a = set(UPRNs_joined_buildings["cluster"].unique())
+b = set(UPRNs_clustered["cluster"].unique())
```

Comment:
```suggestion
a = set(UPRNs_joined_buildings["cluster"])
b = set(UPRNs_clustered["cluster"])
```
very minor but i guess you dont need `.unique()` because it's a set

Reply:
good point!